### PR TITLE
Fix hashing step ids in loops

### PR DESCRIPTION
**Pull request by albertchae**
## Summary

Per https://github.com/inngest/inngest/blob/main/docs/SDK_SPEC.md#512-ids-and-hashing,
add `:n` starting with `:1` for repeated instances of a step id

Refactored to be a combination of golang and inngest-js implementation to handle edge case of user defined stepId colliding
- https://github.com/inngest/inngestgo/blob/0a00daba0b2db68ff0f080f787cf63f0a63b44d8/internal/sdkrequest/manager.go#L123-L131
- https://github.com/inngest/inngest-js/blob/79069e1a3d700624ce49b323922c113fc952bcc6/packages/inngest/src/components/execution/v1.ts#L819-L831

The inngest-js SDK currently optionally warns of parallel indexing, but
this isn't in scope for beta so I left it out.

## Checklist

- [ ] ~~Update documentation~~ N/A documented by sdk spec
- [x] Added unit/integration tests

## Related

- INN-3329

### CODE REVIEW

File: inngest/src/main/kotlin/com/inngest/State.kt [hunk 1]
```diff
@@ -20,6 +29,20 @@ class State(
...
+    private fun findNextAvailableStepId(id: String): String {
+        var stepNumber = 1
+        var possibleStepId: String
+        while (true) {
+            possibleStepId = "$id:$stepNumber"
+            if (possibleStepId !in stepIds) {
+                break
+            }
+            stepNumber++
+        }
```
Comment (albertchae) [inline inngest/src/main/kotlin/com/inngest/State.kt:41]: not sure if there's a better way to find the next available `stepNumber` but looks like this [hasn't been a problem in the inngest-js SDK](https://github.com/inngest/inngest-js/blob/79069e1a3d700624ce49b323922c113fc952bcc6/packages/inngest/src/components/execution/v1.ts#L823-L829) yet so leaving it as is for now

File: inngest-spring-boot-demo/src/main/java/com/inngest/springbootdemo/testfunctions/LoopFunction.java [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+            .triggerEvent("test/loop");
+    }
+
+
+    @Override
+    public Integer execute(FunctionContext ctx, Step step) {
+        int runningCount = 10;
+        for (int i = 0; i < 5; i++) {
+            int effectivelyFinalVariableForLambda = runningCount;
+            runningCount = step.run("add-ten", () -> effectivelyFinalVariableForLambda + 10, Integer.class);
```
Comment (albertchae) [inline inngest-spring-boot-demo/src/main/java/com/inngest/springbootdemo/testfunctions/LoopFunction.java:26]: this would not compile if I tried to use `runningCount` directly in the lambda

File: inngest-spring-boot-demo/src/test/java/com/inngest/springbootdemo/LoopFunctionIntegrationTest.java [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+
+    @Test
+    void testStepsInLoopExecuteCorrectly() throws Exception {
+        String loopEvent = InngestFunctionTestHelpers.sendEvent(client, "test/loop").getIds()[0];
+        Thread.sleep(2000);
+
+        RunEntry<Object> loopRun = devServer.runsByEvent(loopEvent).first();
+        assertEquals("Completed", loopRun.getStatus());
+
+        assertEquals(60, loopRun.getOutput());
```
Comment (albertchae) [inline inngest-spring-boot-demo/src/test/java/com/inngest/springbootdemo/LoopFunctionIntegrationTest.java:28]: was returning 20 before fix to the SDK

File: inngest/src/main/kotlin/com/inngest/State.kt (hunk 1 above)
Comment (darwin67) [reply inngest/src/main/kotlin/com/inngest/State.kt:41]: the Go SDK does something like this.
https://github.com/inngest/inngestgo/blob/0a00daba0b2db68ff0f080f787cf63f0a63b44d8/internal/sdkrequest/manager.go#L123-L131

Comment (darwin67) [reply inngest/src/main/kotlin/com/inngest/State.kt:41]: doing a loop every time to find the next possible one might not be the most efficient. probably having a map or something of the kind to track state temporarily as it goes through the existing state would be ideal

Comment (albertchae) [reply inngest/src/main/kotlin/com/inngest/State.kt]: cool, that's the algorithm I was envisioning anyway. updated to store count in a map

File: inngest/src/main/kotlin/com/inngest/State.kt [hunk 4]
```diff
@@ -20,6 +23,19 @@ class State(
...
 
+    private fun findNextAvailableStepId(id: String): String {
+        if (id !in stepIdsToSeenCount) {
+            stepIdsToSeenCount[id] = 1
+            return id
+        }
+
+        // use the seen count so far for current step, increment for next time
+        val stepNumber = stepIdsToSeenCount[id]
+        stepIdsToSeenCount[id] = stepIdsToSeenCount.getValue(id) + 1
```
Comment (albertchae) [inline inngest/src/main/kotlin/com/inngest/State.kt:34]: `stepIdsToSeenCount[id]++` does not compile because the return value on `[]` map access is `T?`, even with the `!in` check above 😭 

https://blog.danlew.net/2017/06/14/convincing-the-kotlin-compiler-that-code-is-safe/

**Review by KiKoS0 — APPROVED** [review]

File: inngest/src/main/kotlin/com/inngest/State.kt [hunk 5]
```diff
@@ -20,6 +23,19 @@ class State(
...
+
+        return "$id:$stepNumber"
```
Comment (jpwilliams) [inline inngest/src/main/kotlin/com/inngest/State.kt:36]: We may also need a check here to see is this step ID already exists.

A user could do something like:
- Run `"my-step"`
- Run `"my-step:1"` (user explicitly specifying this string)
- Run a loop of `"my-step"` steps

I think here this would result in us redeclaring `"my-step:1"` in the first iteration of the loop even though we'd like to skip it and go straight to `"my-step:2"`.

These IDs just being strings means a user can accidentally stumble into our `[ID]:[count]` format and break some stuff. 😄

Comment (albertchae) [reply inngest/src/main/kotlin/com/inngest/State.kt:46]: OK pushed a commit to combine both a hash for O(1) in most cases and a loop afterwards just in case a user used that stepId already. Correct me if I'm wrong but is this a bug in the Go SDK then https://github.com/inngest/inngestgo/blob/0a00daba0b2db68ff0f080f787cf63f0a63b44d8/internal/sdkrequest/manager.go#L123-L131 @darwin67 ?

This seems like it could potentially be worth reserving some delimiter characters for metadata if Inngest has other cases where it would want to modify the user provided stepId.

**Review by jpwilliams — APPROVED** [review]

File: inngest/src/main/kotlin/com/inngest/State.kt (hunk 5 above)
Comment (albertchae) [reply inngest/src/main/kotlin/com/inngest/State.kt:45]: I thought of another edge case where the user could reuse a `my-step:n` name after a loop that used it, so I updated the test and logic to handle that too
